torch/_torch_docs.py

```diff
@@ -1489,6 +1489,67 @@
         - **v** (*Tensor*): the eigenvectors of ``a`` if ``eigenvectors`` is ``True``; otherwise an empty tensor
 """)
 
+add_docstr(torch.einsum,
+           r"""
+einsum(equation, operands) -> Tensor
+
+This function provides a way of computing multilinear expressions (i.e. sums of products) using the
+Einstein summation convention.
+
+Args:
+    equation (string): The equation is given in terms of lower case letters (indices) to be associated
+           with each dimension of the operands and result. The left hand side lists the operands
+           dimensions, separated by commas. There should be one index letter per tensor dimension.
+           The right hand side follows after `->` and gives the indices for the output.
+           If the `->` and right hand side are omitted, it implicitly defined as the alphabetically
+           sorted list of all indices appearing exactly once in the left hand side.
+           The indices not apprearing in the output are summed over after multiplying the operands
+           entries.
+           `einsum` does not implement diagonals (multiple occurences of a single index for one tensor,
+           e.g. `ii->i`) and ellipses (`...`).
+    operands (list of Tensors): The operands to compute the Einstein sum of.
+           Note that the operands are passed as a list, not as individual arguments.
+
+Examples::
+
+    >>> x = torch.randn(5)
+    >>> y = torch.randn(4)
+    >>> torch.einsum('i,j->ij', (x,y))  # outer product
+
+    -1.0066 -2.0433 -0.8290  0.8429
```

Comment:
Maybe we should regenerate *all* documentation examples regularly (and set a fixed seed before running them)... Then we'd be sure they are up to date and working. I could send a PR for that it is helpful - the documentation is very dear to me.

Reply:
I agree, and it's being discussed in https://github.com/pytorch/pytorch/issues/6662

Reply:
Is there a reason we need to pass arguments as tuples?

Btw great work!

Reply:
Is there a reason we need to pass arguments as tuples?

Great comment, thank you!
The parsing of arguments is in [`torch/csrc/utils/python_arg_parser.cpp`, in `FunctionSignature::parse`](https://github.com/pytorch/pytorch/blob/187955b9599c162f75d19ceb65fde02bd85253b3/torch/csrc/utils/python_arg_parser.cpp#L363). Varargs are only supported if there is a single parameter and that is of type IntList. This is because in the arg parsing [the entire argument tuple is assigned as the argument](https://github.com/pytorch/pytorch/blob/187955b9599c162f75d19ceb65fde02bd85253b3/torch/csrc/utils/python_arg_parser.cpp#L409). Now, if we wanted to have a slice there (which would be the way to support it generically), we would need to construct one using `PyTuple_GetSlice`. But that would be a new reference, and suddenly we would need to release the reference in the indirect caller - who typically used `PythonArgParse.parse` . So the way to support it would be to return something like "vararg-tuple" and deallocate that or have `PythonArgs` do that automatically when it goes out of scope - but I'm not sure whether the latter is a good idea if the caller does have the `ParsedArgs` struct as well. All this seems to be fairly invasive and also I don't understand the rationale of exposing `ParsedArgs` to callers (rather than only showing the `PythonArgs`) to not have that be a fairly strong distraction to the einsum implementation.

The obvious alternative is to do our own arg parsing and blacklist it in automatic generation - similar to what is [done for clamp](https://github.com/pytorch/pytorch/blob/187955b9599c162f75d19ceb65fde02bd85253b3/tools/autograd/templates/python_torch_functions.cpp#L51). The additional complication would be that we cannot use `PythonArgParse` - not that bad if we can still use `PythonArgs`.

As such, I would tentatively not want to burden the reviewers with this. However if they would like to review that along with the other bits, I would be happy to include that in the PR as well.

Reply:
Let's get this as in!